File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    int[] nums;
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        this.nums = nums;
        int max1 = dfs(new Integer[nums.length], 1, nums.length-1);
        int max2 = dfs(new Integer[nums.length], 0, nums.length-2);
        return Math.max(max1, max2) ;
    }
    
    private int dfs(Integer[] memo, int start, int i) {
        if (i < start) {
            return 0;
        }
        if (memo[i] == null) {
            memo[i] = Math.max(dfs(memo, start, i-2)+nums[i], dfs(memo, start, i-1));
        }
        return memo[i];
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) {
            return nums[0];
        }
        this->nums = nums;
        unordered_map<int, int> memo;
        int max1 = dfs(memo, 1, nums.size()-1);
        memo.clear();
        int max2 = dfs(memo, 0, nums.size()-2);
        return max(max1, max2);
    }
    
private: 
    vector<int> nums;
    int dfs(unordered_map<int, int>& memo, int start, int i) {
        if (i < start) {
            return 0;
        }
        if (memo.find(i) == memo.end()) {
            memo[i] = max(dfs(memo, start, i-2)+nums[i], dfs(memo, start, i-1));
        }
        return memo[i];
    }    
};